compiler/packages/babel-plugin-react-compiler/src/Transform/TransformFire.ts

```diff
@@ -0,0 +1,633 @@
+
+/*
+ * TODO(jmbrown):
+ *   In this stack:
+ *     - Insert useFire import
+ *     - Assert no lingering fire calls
+ *     - Ensure a fired function is not called regularly elsewhere in the same effect
+ *
+ *   Future:
+ *   - rewrite dep arrays
+ *   - traverse object methods
+ *   - method calls
+ */
+
+const CANNOT_COMPILE_FIRE = 'Cannot compile `fire`';
+
+function deleteInstructions(
+  deleteInstrs: Set<InstructionId>,
+  instructions: Array<Instruction>,
+): Array<Instruction> {
+  if (deleteInstrs.size > 0) {
+    const newInstrs = instructions.filter(instr => !deleteInstrs.has(instr.id));
+    return newInstrs;
+  }
+  return instructions;
+}
+
+function rewriteInstructions(
+  rewriteInstrs: Map<InstructionId, Array<Instruction>>,
+  instructions: Array<Instruction>,
+): Array<Instruction> {
+  if (rewriteInstrs.size > 0) {
+    const newInstrs = [];
+    for (const instr of instructions) {
+      const newInstrsAtId = rewriteInstrs.get(instr.id);
+      if (newInstrsAtId != null) {
+        newInstrs.push(...newInstrsAtId, instr);
+      } else {
+        newInstrs.push(instr);
+      }
+    }
+
+    return newInstrs;
+  }
+
+  return instructions;
+}
+
+function makeLoadUseFireInstruction(env: Environment): {
+  loadUseFireInstr: Instruction;
+  useFirePlace: Place;
+} {
+  const useFirePlace = createTemporaryPlace(env, GeneratedSource);
+  useFirePlace.effect = Effect.Read;
+  useFirePlace.identifier.type = DefaultNonmutatingHook;
+  const instrValue: InstructionValue = {
+    kind: 'LoadGlobal',
+    binding: {
+      kind: 'Global',
+      name: 'useFire',
+    },
+    loc: GeneratedSource,
+  };
+  return {
+    loadUseFireInstr: {
+      id: makeInstructionId(0),
+      value: instrValue,
+      lvalue: useFirePlace,
+      loc: GeneratedSource,
+    },
+    useFirePlace,
+  };
+}
+
+function makeLoadFireCalleeInstruction(
+  env: Environment,
+  fireCalleeIdentifier: Identifier,
+): {loadFireCalleeInstr: Instruction; loadedFireCallee: Place} {
+  const loadedFireCallee = createTemporaryPlace(env, GeneratedSource);
+  const fireCallee: Place = {
+    kind: 'Identifier',
+    identifier: fireCalleeIdentifier,
+    reactive: false,
+    effect: Effect.Unknown,
+    loc: fireCalleeIdentifier.loc,
+  };
+  return {
+    loadFireCalleeInstr: {
+      id: makeInstructionId(0),
+      value: {
+        kind: 'LoadLocal',
+        loc: GeneratedSource,
+        place: fireCallee,
+      },
+      lvalue: loadedFireCallee,
+      loc: GeneratedSource,
+    },
+    loadedFireCallee,
+  };
+}
+
+function makeCallUseFireInstruction(
+  env: Environment,
+  useFirePlace: Place,
+  argPlace: Place,
+): {callUseFireInstr: Instruction; useFireCallResultPlace: Place} {
+  const useFireCallResultPlace = createTemporaryPlace(env, GeneratedSource);
+  useFireCallResultPlace.effect = Effect.Read;
+
+  const useFireCall: CallExpression = {
+    kind: 'CallExpression',
+    callee: useFirePlace,
+    args: [argPlace],
```

Comment:
these places should be spread - as a general rule, always spread every place where you use it in an HIR value

Reply:
(the reason is that the same instance is also used as lvalue in a preceding instruction)

Reply:
Applied this general rule to all HIR instructions in the diff